File: web/js/engine.js
// The "follow the winners" (momentum) backtest engine — JavaScript port of
// tradelab/engine.py + the momentum strategy, with the same honesty rules:
//   * decisions use a day's CLOSE, orders execute at the NEXT day's OPEN
//   * every fill pays slippage
//   * safety exits check the day's actual LOW; gaps fill at the open
//   * new buys are blocked while the S&P 500 is below its 200-day average

export function smaArr(values, window) {
  const out = new Array(values.length).fill(null);
  let sum = 0;
  for (let i = 0; i < values.length; i++) {
    sum += values[i];
    if (i >= window) sum -= values[i - window];
    if (i >= window - 1) out[i] = sum / window;
  }
  return out;
}

export function runBacktest(cfg, data, spy) {
  const {
    start, end, capital,
    lookbackDays, skipDays, topN, holdRank, rebalanceDays,
    maxPositions, positionPct, stopLossPct, trailingStopPct,
    regimeFilter, slippageBps,
  } = cfg;
  const slip = slippageBps / 10000;

  // Trading calendar = SPY dates inside the window.
  const cal = [];
  for (let i = 0; i < spy.dates.length; i++) {
    if (spy.dates[i] >= start && spy.dates[i] <= end) cal.push(i);
  }
  if (cal.length < 10) throw new Error("Backtest window too short / no benchmark data.");
  const spySma = smaArr(spy.c, 200);

  const bar = (t, d) => {
    const td = data.get(t);
    if (!td) return null;
    const i = td.idx.get(d);
    return i == null ? null : { o: td.o[i], h: td.h[i], l: td.l[i], c: td.c[i] };
  };

  const rankAll = (d) => {
    const scored = [];
    for (const [t, td] of data) {
      const i = td.idx.get(d);
      if (i == null) continue;
      const j = i - skipDays, k = i - skipDays - lookbackDays;
      if (k < 0) continue;
      scored.push([t, td.c[j] / td.c[k] - 1]);
    }
    scored.sort((a, b) => b[1] - a[1]);
    const ranks = new Map();
    scored.forEach(([t], i) => ranks.set(t, i + 1));
    return ranks;
  };

  let cash = capital;
  const positions = new Map();
  const trades = [];
  const equityDates = [], equityVals = [];
  let pendingBuys = [];
  const pendingSells = new Map();
  let lastRebalIdx = null;
  let currentRanks = null;

  const closePos = (t, d, price, reason) => {
    const p = positions.get(t);
    positions.delete(t);
    const fill = price * (1 - slip);
    cash += p.shares * fill;
    const tr = trades.find((x) => x.ticker === t && x.exitDate === null);
    if (tr) {
      tr.exitDate = d;
      tr.exitPrice = fill;
      tr.pnl = (fill - tr.entryPrice) * tr.shares;
      tr.pnlPct = fill / tr.entryPrice - 1;
      tr.reason = reason;
    }
  };

  for (let k = 0; k < cal.length; k++) {
    const spyI = cal[k];
    const d = spy.dates[spyI];

    // 1. Execute yesterday's sell decisions at today's open
    for (const [t, reason] of pendingSells) {
      const b = bar(t, d);
      if (b && positions.has(t)) closePos(t, d, b.o, reason);
    }
    pendingSells.clear();

    // ...then yesterday's buy decisions
    for (const t of pendingBuys) {
      if (positions.has(t) || positions.size >= maxPositions) continue;
      const b = bar(t, d);
      if (!b) continue;
      let port = cash;
      for (const [pt, p] of positions) {
        const pb = bar(pt, d);
        port += p.shares * (pb ? pb.c : p.highWater);
      }
      const budget = Math.min((port * positionPct) / 100, cash);
      const fill = b.o * (1 + slip);
      const shares = budget / fill;
      if (shares <= 0 || shares * fill > cash + 1e-9) continue;
      cash -= shares * fill;
      positions.set(t, { entryDate: d, entryPrice: fill, shares, highWater: fill, daysHeld: 0 });
      trades.push({ ticker: t, entryDate: d, entryPrice: fill, shares, exitDate: null, exitPrice: null, pnl: null, pnlPct: null, reason: "" });
    }
    pendingBuys = [];

    // 2. Safety exits — check stops against today's actual low
    for (const [t, p] of [...positions]) {
      const b = bar(t, d);
      if (!b) continue;
      const stops = [];
      if (stopLossPct > 0) stops.push(p.entryPrice * (1 - stopLossPct / 100));
      if (trailingStopPct > 0) stops.push(p.highWater * (1 - trailingStopPct / 100));
      if (!stops.length) continue;
      const stop = Math.max(...stops);
      if (b.l <= stop) closePos(t, d, Math.min(b.o, stop), `safety exit at $${stop.toFixed(2)}`);
    }

    // 3. End of day: mark to market, update high-water marks
    let port = cash;
    for (const [t, p] of positions) {
      const b = bar(t, d);
      if (b) {
        p.highWater = Math.max(p.highWater, b.c);
        p.daysHeld++;
        port += p.shares * b.c;
      } else {
        port += p.shares * p.highWater;
      }
    }
    equityDates.push(d);
    equityVals.push(port);

    // Strategy exits: holdings that fell down the strength ranking
    if (currentRanks) {
      for (const [t] of positions) {
        const r = currentRanks.get(t);
        if (r == null || r > holdRank) {
          pendingSells.set(t, `dropped to #${r ?? "–"} in the strength ranking (limit ${holdRank})`);
        }
      }
    }

    // Strategy entries: monthly re-rank, only in a healthy market
    const regimeOK = !regimeFilter || (spySma[spyI] != null && spy.c[spyI] > spySma[spyI]);
    if (regimeOK) {
      const slots = maxPositions - (positions.size - pendingSells.size);
      const due = lastRebalIdx === null || k - lastRebalIdx >= rebalanceDays;
      if (slots > 0 && due) {
        lastRebalIdx = k;
        currentRanks = rankAll(d);
        const cands = [...currentRanks.entries()]
          .filter(([, r]) => r <= topN)
          .sort((a, b) => a[1] - b[1])
          .map(([t]) => t)
          .filter((t) => !positions.has(t) && !pendingSells.has(t));
        pendingBuys = cands.slice(0, slots);
      }
    }
  }

  // Benchmark: buy SPY at the first day's open with the same money + slippage
  const firstI = cal[0];
  const spyFill = spy.o[firstI] * (1 + slip);
  const benchVals = cal.map((i) => (spy.c[i] * capital) / spyFill);

  const warnings = [];
  const closed = trades.filter((t) => t.exitDate !== null);
  if (closed.length < 30) {
    warnings.push(`Only ${closed.length} finished trades — that's a small sample; ` +
      "test a longer period before trusting the result.");
  }
  warnings.push("The stock list is today's giants — companies that collapsed in the " +
    "past aren't included, so multi-year results look rosier than reality.");

  return { equityDates, equityVals, benchVals, trades, warnings };
}

File: tradelab/engine.py
"""The backtest engine.

Correctness rules baked in:
  * Signals use data up to day T's close; orders execute at day T+1's OPEN.
  * Every fill pays slippage (default 5 bps) — no free lunches.
  * Stops are checked against the day's actual low; gap-downs fill at the
    open, not at the stop price (the market doesn't honor your stop level).
  * Cash can never go negative; position sizes are % of *current* equity.

The engine owns portfolio/risk logic (sizing, stops, regime filter,
max positions); strategies only emit entry/exit wishes.
"""

from dataclasses import dataclass, field

import pandas as pd

from . import indicators as ind
from .strategies import Position, STRATEGIES


@dataclass
class Trade:
    ticker: str
    entry_date: pd.Timestamp
    entry_price: float
    shares: float
    exit_date: pd.Timestamp = None
    exit_price: float = None
    pnl: float = None
    pnl_pct: float = None
    reason: str = ""


@dataclass
class BacktestResult:
    equity: pd.Series = None            # daily equity curve
    benchmark: pd.Series = None         # SPY buy-and-hold, same capital
    trades: list = field(default_factory=list)
    config: dict = field(default_factory=dict)
    warnings: list = field(default_factory=list)


DEFAULT_RISK = {
    "max_positions": 5,
    "position_pct": 20.0,        # % of current equity per new position
    "stop_loss_pct": 15.0,       # fixed stop below entry; 0 disables
    "trailing_stop_pct": 20.0,   # trail below highest close; 0 disables
    "max_hold_days": 0,          # 0 disables
    "regime_filter": True,       # block new entries when SPY < 200-day SMA
    "regime_sma": 200,
}

DEFAULT_COSTS = {"slippage_bps": 5.0, "commission": 0.0}


def run_backtest(
    strategy_name: str,
    strategy_params: dict,
    data: dict,                 # {ticker: OHLCV df} incl. warmup history
    benchmark_df: pd.DataFrame, # SPY OHLCV
    start: str,
    end: str,
    capital: float = 1000.0,
    risk: dict = None,
    costs: dict = None,
    fractional_shares: bool = True,
) -> BacktestResult:
    risk = {**DEFAULT_RISK, **(risk or {})}
    costs = {**DEFAULT_COSTS, **(costs or {})}
    slip = costs["slippage_bps"] / 10_000.0
    commission = costs["commission"]

    strat_cls = STRATEGIES[strategy_name]
    strat = strat_cls(strategy_params, data)

    # Trading calendar = benchmark dates inside the window.
    cal = benchmark_df.index
    cal = cal[(cal >= pd.Timestamp(start)) & (cal <= pd.Timestamp(end))]
    if len(cal) < 10:
        raise ValueError("Backtest window too short or no benchmark data.")

    regime_ok = pd.Series(True, index=benchmark_df.index)
    if risk["regime_filter"]:
        regime_ok = benchmark_df["Close"] > ind.sma(
            benchmark_df["Close"], int(risk["regime_sma"])
        )

    cash = capital
    positions: dict[str, Position] = {}
    trades: list[Trade] = []
    equity_curve = {}
    # Orders decided on yesterday's close, to execute at today's open.
    pending_buys: list[str] = []
    pending_sells: dict[str, str] = {}  # ticker -> reason

    def bar(ticker, date):
        df = data.get(ticker)
        if df is None or date not in df.index:
            return None
        return df.loc[date]

    def close_position(ticker, date, price, reason):
        nonlocal cash
        pos = positions.pop(ticker)
        fill = price * (1 - slip)
        cash += pos.shares * fill - commission
        for t in trades:
            if t.ticker == ticker and t.exit_date is None:
                t.exit_date = date
                t.exit_price = fill
                t.pnl = (fill - t.entry_price) * t.shares
                t.pnl_pct = fill / t.entry_price - 1
                t.reason = reason
                break

    for date in cal:
        # ---- 1. Execute yesterday's decisions at today's open ----
        for ticker, reason in list(pending_sells.items()):
            b = bar(ticker, date)
            if b is not None and ticker in positions:
                close_position(ticker, date, b["Open"], reason)
        pending_sells.clear()

        for ticker in pending_buys:
            if ticker in positions or len(positions) >= int(risk["max_positions"]):
                continue
            b = bar(ticker, date)
            if b is None:
                continue
            port_value = cash + sum(
                p.shares * (bar(t, date)["Close"] if bar(t, date) is not None else p.high_water)
                for t, p in positions.items()
            )
            budget = min(port_value * risk["position_pct"] / 100.0, cash)
            fill = b["Open"] * (1 + slip)
            shares = budget / fill if fractional_shares else int(budget // fill)
            if shares <= 0 or shares * fill > cash:
                continue
            cash -= shares * fill + commission
            positions[ticker] = Position(ticker, date, fill, shares)
            trades.append(Trade(ticker, date, fill, shares))
        pending_buys = []

        # ---- 2. Intraday stop checks against today's low ----
        for ticker, pos in list(positions.items()):
            b = bar(ticker, date)
            if b is None:
                continue
            stop_levels = []
            if risk["stop_loss_pct"] > 0:
                stop_levels.append(pos.entry_price * (1 - risk["stop_loss_pct"] / 100))
            if risk["trailing_stop_pct"] > 0:
                stop_levels.append(pos.high_water * (1 - risk["trailing_stop_pct"] / 100))
            if not stop_levels:
                continue
            stop = max(stop_levels)
            if b["Low"] <= stop:
                # Gap through the stop fills at the open, not the stop.
                fill_price = min(b["Open"], stop)
                close_position(ticker, date, fill_price, f"stop hit at {stop:.2f}")

        # ---- 3. End of day: mark to market, update trails, decide ----
        port_value = cash
        for ticker, pos in positions.items():
            b = bar(ticker, date)
            if b is not None:
                pos.high_water = max(pos.high_water, b["Close"])
                pos.days_held += 1
                port_value += pos.shares * b["Close"]
            else:
                port_value += pos.shares * pos.high_water
        equity_curve[date] = port_value

        # Strategy exit wishes -> sell at tomorrow's open
        for ticker, pos in positions.items():
            reason = strat.wants_exit(ticker, date, pos)
            if reason is None and risk["max_hold_days"] and pos.days_held >= int(risk["max_hold_days"]):
                reason = f"max hold {risk['max_hold_days']} days"
            if reason:
                pending_sells[ticker] = reason

        # Strategy entry wishes -> buy at tomorrow's open (regime permitting)
        allowed = bool(regime_ok.get(date, True))
        if allowed:
            slots = int(risk["max_positions"]) - (len(positions) - len(pending_sells))
            if slots > 0:
                cands = [
                    t for t in strat.entry_candidates(date)
                    if t not in positions and t not in pending_sells
                ]
                pending_buys = cands[:slots]

    equity = pd.Series(equity_curve).sort_index()

    # Benchmark: buy SPY at the first day's open with the same capital+slippage.
    bench_window = benchmark_df.loc[equity.index[0]: equity.index[-1]]
    spy_fill = bench_window["Open"].iloc[0] * (1 + slip)
    benchmark = bench_window["Close"] * (capital / spy_fill)

    warnings = []
    closed = [t for t in trades if t.exit_date is not None]
    if len(closed) < 30:
        warnings.append(
            f"Only {len(closed)} closed trades — results are statistically weak; "
            "test a longer period before trusting them."
        )
    warnings.append(
        "Universe is today's large caps (survivorship bias) — treat multi-year "
        "results as optimistic."
    )

    return BacktestResult(
        equity=equity,
        benchmark=benchmark,
        trades=trades,
        config={
            "strategy": strategy_name, "params": strategy_params,
            "risk": risk, "costs": costs, "capital": capital,
            "start": str(equity.index[0].date()), "end": str(equity.index[-1].date()),
            "fractional_shares": fractional_shares,
        },
        warnings=warnings,
    )
